### DISCUSSION TITLE
How to use storageBuffer in a vertexShader without a pipeline's layout defined to "auto" ?

### DISCUSSION THREAD

**Discussion by tlecoz** [gpuweb/gpuweb#4147] [Q&A] [answered]
If you try to use an uniformBuffer as a storageBuffer in a vertexShader , it works without any problem if you use the combo 

`pipeline.layout = "auto"; `

and 

```
bindGroup = device.createBindGroup({
    layout: pipeline.getBindGroupLayout(0),
```
    

But if you try to do the same thing with a well defined bindgroupLayout , you get this error : 

"buffer bindings cannot be used with a visibility of ShaderStage::(Vertex|Fragment). Only ShaderStage::(Fragment|Compute) are allowed.
 - While validating entries[0]
 - While validating [BindGroupLayoutDescriptor]
 - While calling [Device].CreateBindGroupLayout([BindGroupLayoutDescriptor])." 
 
How can I reproduce the magic begind "pipeline.getBindGroupLayout(0)" that allow the storage buffer to be used in a vertexShader with a well defined bindgroupLayout ?

**Comment by Kangz** [comment, marked as answer]
You need to use the [`read-only-storage`](https://gpuweb.github.io/gpuweb/#dom-gpubufferbindingtype-read-only-storage) buffer binding type in your bind group layout. That's what the implicit layout should be doing in your case.

**Comment by tlecoz** [comment]
Thank you a lot ! It works !
